text/0328-polyglot-assert.md

```diff
@@ -0,0 +1,335 @@
+---
+rfc pr: https://github.com/aws/aws-cdk-rfcs/pull/330
+tracking issue: https://github.com/aws/aws-cdk-rfcs/issues/328
+---
+
+# Polyglot Assert
+
+The `@aws-cdk/assert` module is only available to javascript users.
+
+The new polyglot assert module - `@aws-cdk/assertv2` - provides the same set of functionalities
+to users in all CDK supported languages.
+
+## Working Backwards
+
+### CHANGELOG
+
+**feat:** announcing assertv2: the assert library is now available in all CDK supported languages.
+
+### README
+
+<!--BEGIN STABILITY BANNER-->
+
+---
+
+![cdk-constructs: Stable](https://img.shields.io/badge/cdk--constructs-stable-success.svg?style=for-the-badge)
+
+---
+
+<!--END STABILITY BANNER-->
+
+> NOTE: This module contains *beta APIs*.
+>
+> All beta symbols are suffixed with the `Beta<n>`. When we have backwards
+> incompatible change, we will create a new symbol with a `Beta<n+1>` suffix
+> and deprecate the `Beta<n>` symbol.
+---
+This module allows asserting the contents of CloudFormation templates.
+
+To run assertions based on a CDK `Stack`, start off with -
+
+```ts
+const stack = new cdk.Stack(...)
+...
+const inspect = StackAssertionsBeta1.fromStackBeta1(stack);
+```
+
+Alternatively, assertions can be run on an existing CloudFormation template -
+
+```ts
+const file = '/path/to/file';
+const inspect = StackAssertionsBeta1.fromTemplateFileBeta1(file);
+```
+
+#### Full Template Match
+
+The simplest assertion would be to assert that the template matches a given
+template.
+
+```ts
+inspect.assertMatchTemplateBeta1({
+  Resources: {
+    Type: 'Foo::Bar',
+    Properties: {
+      Baz: 'Qux',
+    },
+  },
+});
+```
+
+#### Counting Resources
+
+This module allows asserting the number of resources of a specific type found
+in a template.
+
+```ts
+inspect.assertResourceCountBeta1('Foo::Bar', 2);
+```
+
+#### Resource Matching
+
+Beyond resource counting, the module also allows asserting that a resource with
+specific properties are present.
+
+The following code asserts that the `Properties` section of a resource of type
+`Foo::Bar` contains the specified properties -
+
+```ts
+inspect.assertResourceBeta1('Foo::Bar', {
+  Foo: 'Bar',
+  Baz: 5,
+  Qux: [ 'Waldo', 'Fred' ],
+});
+```
+
+The same method allows asserting the complete definition of the 'Resource'
+which can be used to verify things other sections like `DependsOn`, `Metadata`,
+`DeletionProperty`, etc.
+
+```ts
+inspect.assertResourceBeta1('Foo::Bar', {
+  Properties: { Foo: 'Bar' },
+  DependsOn: [ 'Waldo', 'Fred' ],
+}, {
+  part: ResourcePartBeta1.COMPLETE_BETA1,
```

Comment:
Again, no need to add _BETA1 everywhere. It's sufficient that the top level type has that prefix...

Reply:
Let's say I wanted to change the behaviour of this enum value. With your suggestion I would need to create a new enum `ResourcePartBeta2` and then create new methods with `Beta2` (or whatever) suffix everywhere this enum is used.

More pragmatic to keep the suffix, no?

Reply:
I am not sure I can think of a reason you'd want to change the value of the enum member. It's normally not even needed to specify a value.

We should be consistent about these suffixes. If the entire type is "Beta", I don't see any value in actually polluting all the members with that suffix.

Reply:
To @nija-at point, if we only use `betaX` on the containing type (e.g class), what do we do when we want to deprecate an internal member. Fro example, lets say the `autoScalingGroup` class is experimental, and we only want to change the behavior of `formAutScalingGroupName` method, do we need to deprecate the entire `autoScalingGroup` class for that?


```
export class AutoScalingGroupBeta  {

   public static fromAutoScalingGroupName(scope: Construct, id: string, autoScalingGroupName: string): 
    .....
  }
}
```  


My  gut feeling is that  it will be  less  ideal from a  user experience  prespecitve. 

there  are a lot  of nuance here, I think we need to flush out the preview specification so we can be consistent across the code base.

Reply:
I would just deprecate `AutoScalingGroupBeta`, and introduce `AutoScalingGroupBeta2` with this new method, while proxying all the deprecated methods from the deprecated type to the new type.

Reply:
How will this work for users consuming `fromAutoScalingGroupName`? If you proxy it, you will be introducing a breaking change

Reply:
Good callout. Of course this specific method will not be proxied as-is in order to preserve the legacy behavior in the deprecated class.

Reply:
Agree with Elad here. Having to call `fromAutoScalingGroupName()` `AutoScalingGroupBeta1.fromAutoScalingGroupNameBeta1()` on the off-chance we will want to deprecate it seems like the wrong tradeoff to me.

Also, let's assume for a second we do go down this route, and call it `AutoScalingGroupBeta1.fromAutoScalingGroupNameBeta1()`. What do we call the replacement method then? `AutoScalingGroupBeta1.fromAutoScalingGroupNameBeta2()`? That looks pretty bad...

Reply:
I think we are over indexing on how things look. Our main objective is to make sure it is clear that an API is in preview. The second objective is to make is easy for users to use preview APIs, will introducing a whole new class will make it harder for users to upgrade? It can be, as the class usage is more prevalent than a single method in a user application (more references to the `AutoScalingGroup` class in the code base than to `fromAutoScalingGroupName`).